File: kmeans.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
#include <float.h>
#include <ctype.h>

typedef struct list ELEMENT;
typedef ELEMENT *LINK;
typedef double *datapoint; 
/* LINK is a pointer to ELEMENT */
/* datapoint is a pointer to a double array */

void anErrorHasOccurred();
void invalidInput();
double** fileToDataPoints(FILE *ifp,int d, int size);
int findDimension(FILE *ifp);
int findInputSize(FILE *ifp);
double **inizializeCentroids(int k, int d, double **datapoints);
void restartClusters(LINK *clusters, int k);
void delete_list(LINK head);
void kMeans(int k, int size, int d, int max_iter, double **cents, LINK *clusters, double **matrix);
void assignToCluster(double **cents,double** datapoints,LINK *clusters, int k, int size, int d);
int updateCentroids(double **cents, LINK *clusters, double** inputMatrix ,int k, int d);
double calculateNorma(double *old, double * new, int d);
double calculateDistance(double *datapoint, double *centroid, int d);
int isNatural(char * s);
void writeToFile(double **cents, int k, int d, char * file_name);
void validityCheck1(int argc, char *argv[]);
void validityCheck2(int k, int max_iter, int size);

struct list { 
    int datapoint; 
    struct list *next;   
};

void anErrorHasOccurred(){
    printf("An Error Has Occurred");
    exit(1);
}

void invalidInput(){
    printf("Invalid Input!");
    exit(1);
}

double ** fileToDataPoints(FILE *ifp, int d, int size) {
    double currentCoordinate;
    double *vector = NULL;
    double **matrix = NULL;
    int i, j;
    
    vector = (double *)calloc(size*d, sizeof(double));
    if (vector == NULL){
        anErrorHasOccurred();
    }
    matrix = (double **)calloc(size, sizeof(double *)); 
    if (matrix == NULL){
        anErrorHasOccurred();
    }
    for (i=0; i<size; i++) {
        matrix[i] = vector + i*d;
    }
    for (i=0; i<size; i++){  
        for(j=0; j<d; j++){
            fscanf(ifp, "%lf %*[,]", &currentCoordinate);
            matrix[i][j] = currentCoordinate;
        }
    }
    return matrix;
    }

int findDimension(FILE *ifp) {
    char c;
    int commacnt=0;
    int dim=0;
    do {
        c = fgetc(ifp);
        /* Taking input single character at a time*/
        if( feof(ifp) ){
            return -1; 
        }

        if (c == 10) /* c==\n */
            {
                dim = commacnt+1;
                break;
            }
        else if (c==44){
            /* read first line #of, +1 = d, go to the beggining and create array of arrays (d*sizeof(double))
               if digit , dot or - => add to the corrent cordinate of (X_i)
               else if , => add to */
                commacnt++;
            } 
        } while(1); 
    
    rewind(ifp);
    return dim;
}

int findInputSize(FILE *ifp) {
    char c;
    int cnt = 0;
    do {
        c = fgetc(ifp);
        /* Taking input single character at a time*/
        if( feof(ifp) ){
            break; 
        }

        if (c == 10) /* c==\n */ 
            {
                cnt++;
            }
        } while(1); 
    
    rewind(ifp);
    return cnt;
}

double ** inizializeCentroids(int k, int d, double **datapoints){
    double *centroid = NULL;
    double **matrix = NULL;
    int i,j;
    centroid = (double *)calloc(k*d, sizeof(double));
    if (centroid == NULL){
        anErrorHasOccurred();
    }
    matrix = (double **)calloc(k, sizeof(double *)); 
    if (matrix == NULL){
        anErrorHasOccurred();
    }
    for (i=0; i<k; i++) {
        matrix[i] = centroid + i*d;
    }
    for (i=0; i<k; i++){
        for (j=0; j<d; j++){
            matrix[i][j] = datapoints[i][j]; 
        }
    }  
    return matrix; 
}

void restartClusters(LINK *clusters, int k) {
    int i;
    LINK current;
    for (i=0; i<k; i++) {
        current = clusters[i];
        delete_list(current);
        clusters[i] = (ELEMENT*)malloc(sizeof(ELEMENT));
        if (clusters[i] == NULL){
            anErrorHasOccurred();
        }
        clusters[i]->datapoint = -1;
        clusters[i]->next = NULL;
    }
}

void delete_list(LINK head) {
    if (head != NULL){
        delete_list(head->next);
        free(head);
    }
}

void kMeans(int k, int size, int d, int max_iter, double **cents, LINK *clusters, double **matrix){
    int iter = 0;
    int continue_condition = 1;
    while (iter<max_iter && continue_condition){
    
       assignToCluster(cents, matrix, clusters, k, size, d);
       continue_condition = updateCentroids(cents, clusters, matrix, k, d);
       iter++;
    }
}

void assignToCluster(double **cents,double** datapoints,LINK *clusters, int k, int size, int d) {
    double distance;
    double tempDist = 0;
    int clusterIndex, i, j;
    LINK current, new;
 
    for(i=0; i<size; i++) {
        clusterIndex = -1;
        distance = DBL_MAX;
        for (j=0; j<k; j++){
            tempDist = calculateDistance(datapoints[i], cents[j], d);
            if (tempDist < distance) {
                distance = tempDist;
                clusterIndex = j;
            }
        }
        if (clusters[clusterIndex]->datapoint == -1) {
            clusters[clusterIndex]->datapoint = i;
        }else {
            current = clusters[clusterIndex];
            new = (ELEMENT*)malloc( sizeof(ELEMENT));
            if (new == NULL){
                anErrorHasOccurred();
            }
            new -> datapoint = i;
            new -> next = current;
            clusters[clusterIndex] = new;
        }
    }
}

int updateCentroids(double **cents, LINK *clusters, double** inputMatrix ,int k, int d) {
    int i, j, m, s, difference = 0, sizeOfCluster;
    double epsilon = 0.001;
    LINK current = NULL;
    double *sum = (double *)calloc(d, sizeof(double));
    if (sum == NULL){
        anErrorHasOccurred();
    }
    for (i=0; i<k; i++) {
        current = clusters[i];
        sizeOfCluster = 0;
        while (current != NULL && current->datapoint != -1){
            for  (j=0; j<d; j++) {
                sum[j] += inputMatrix[current->datapoint][j];
            }

            current = current->next;
            sizeOfCluster++;
        }
        if (sizeOfCluster == 0){
            anErrorHasOccurred();
        }

        for (m=0; m<d; m++){
            sum[m] /= sizeOfCluster;
        }
        difference += (calculateNorma(cents[i], sum, d) > epsilon)?1:0;
        for (s=0; s<d; s++) {
            cents[i][s] = sum[s];
            sum[s] = 0;
        }
    }
    restartClusters(clusters, k);
    free(sum);
    return difference > 0;
}

double calculateNorma(double *old, double * new, int d){
    double sum = 0;
    int i;
    for (i=0; i<d; i++){
        sum += pow(old[i]-new[i],2);
    }
    return sqrt(sum);
}

double calculateDistance(double *datapoint, double *centroid, int d){
    double distance = 0;
    int j;
    for (j = 0; j<d; j++){
        distance+= pow(datapoint[j] - centroid[j],2);
    }
    return distance;
}

int isNatural(char *s){
    while (*s) {
        if(isdigit(*s) == 0){
            return 0;
        }
        s++;
    }
    return 1;
}

void writeToFile(double **cents, int k, int d, char * fileName){
    int i,j;
    FILE *ofp = NULL;
    ofp = fopen(fileName,"w");
    if (ofp == NULL){
        invalidInput();
    }
    for (i = 0; i<k; i++){
        for (j=0; j<d-1; j++){
            fprintf(ofp,"%.4f,",cents[i][j]);
        }
        fprintf(ofp,"%.4f",cents[i][j]);
        fprintf(ofp,"\n");
    }
    fclose(ofp);
}

void validityCheck1(int argc, char *argv[]){
    if ((argc != 4 && argc != 5) || 
        !isNatural(argv[1]) || 
        (argc == 5 && !isNatural(argv[2]))){
        printf("Invalid Input!");
        exit(1);
        }
}
void validityCheck2(int k, int max_iter, int size){
    if (k == 0 ||
        max_iter == 0 ||
        k>size){
        printf("Invalid Input!");
        exit(1);
        }
}

int main(int argc, char *argv[]){
    int d, size, k, max_iter = 200;
    char *inputFileName = NULL;
    char *outputFileName = NULL;
    FILE *ifp=NULL;
    double **datapointMatrix, **centroids;
    LINK *clusters;
    
    validityCheck1(argc,argv);
    
    k = atoi(argv[1]);

    if (argc==5){
        max_iter = atoi(argv[2]);
        inputFileName=argv[3];
        outputFileName=argv[4];
    }
    else if (argc==4){
        inputFileName=argv[2];
        outputFileName=argv[3];
    }
    
    ifp = fopen(inputFileName, "r");
    if (ifp == NULL){
        invalidInput();
    }
    d = findDimension(ifp);
    size = findInputSize(ifp);
    validityCheck2(k, max_iter, size);
    datapointMatrix = fileToDataPoints(ifp, d, size);
    fclose(ifp);
    
    centroids = inizializeCentroids(k, d, datapointMatrix); /*this array holds K datapoints*/
    clusters = (LINK *)calloc(k, sizeof(LINK));
    if (clusters == NULL){
        anErrorHasOccurred();
    }
    restartClusters(clusters, k);
    kMeans(k, size, d, max_iter, centroids, clusters, datapointMatrix);
    writeToFile(centroids, k, d, outputFileName);
    free(datapointMatrix);
    free(centroids);
    free(clusters);
    return 0;
}

File: kmeans.py
import sys 
import math 

def an_error_has_occurred():
    sys.exit("An Error Has Occurred")

def files_to_dataframe(file_name_1,file_name_2):
    file1 = pd.read_csv(file_name_1)
    size1=file1.shape[1]
    file2 = pd.read_csv(file_name_1)
    size = [str(i) for i in range(file1.shape[1])]
    file1 = pd.read_csv(file_name_1, names=size)
    size = [str(i+size1-1) for i in range(file2.shape[1])]
    size[0] = str(0)
    file2 = pd.read_csv(file_name_2, names=size)

    data = pd.merge(file1, file2, on ='0')
    data=data.drop(['0'],axis=1)
def create_centroids(k, input_mat):
    centroids=[]
    centroids=[input_mat[i] for i in range(k)]
    return centroids

def calculate_distance(centroid, data_point):
    return sum([pow(centroid[i]-data_point[i],2) for i in range(len(centroid))])

def calculate_norma(old_centroid, new_centroid):
    return math.sqrt(sum([pow(old_centroid[i]-new_centroid[i],2) for i in range(len(old_centroid))]))

def assign_to_clusters(input_mat,centroids,k):
    clusters=[[]]*k
    for idx,data_point in enumerate(input_mat):
        distance= float("inf")
        temp_dist=0
        cluster_index=0
        for i, centroid in enumerate(centroids):
            temp_dist=calculate_distance(centroid, data_point)
            if (temp_dist<distance):
                distance = temp_dist
                cluster_index = i
        if (clusters[cluster_index] == []):
            clusters[cluster_index]=[idx]
        else: 
            clusters[cluster_index].append(idx)
    return clusters


def update_centroids(input_mat,centroids,clusters):
    try:
        d=len(input_mat[0])
        epsilon=0.001
        diff=0
        for i ,current_cluster in enumerate(clusters):
            sum_of_vectors=[0]*d
            size_of_cluster = len(current_cluster)
            for idx in current_cluster:
                for j in range(d):
                    sum_of_vectors[j]+=input_mat[idx][j] 
            sum_of_vectors=[float(sum_of_vectors[m]/size_of_cluster) for m in range(d)]
            diff += 1 if calculate_norma(centroids[i],sum_of_vectors)>epsilon else 0
            centroids[i]=sum_of_vectors[:]
           
        return diff>0
    except (ZeroDivisionError): #TypeError
        an_error_has_occurred()

def write_to_file(file_name, centroids):
    try:
        f=open(file_name,"w")
        for centroid in centroids:
            res=""
            for indx in (range(len(centroid)-1)):
                res+="{:.4f}".format(centroid[indx])+","
            res+="{:.4f}".format(centroid[-1])+"\n"
            f.write(res)
        f.close()
    except IOError: 
        an_error_has_occurred()

def kmeans(input_mat , k ,max_iter):
    
    centroids=create_centroids(k,input_mat)
    iter=0
    next_iteration=True
    while(iter<max_iter and next_iteration):        
        clusters=assign_to_clusters(input_mat,centroids,k)
        next_iteration=update_centroids(input_mat,centroids,clusters)
        iter+=1
    return centroids

def invalid_input():
    sys.exit("Invalid Input!")

def check_is_natural(num):
    try:
        # Convert it into float
        
        val_f = float(num)
        val_int=int(float(num))
    
    except ValueError:
        invalid_input()
    
    if val_f!=val_int or val_int<=0:
        invalid_input()



def main(): 
    try:
        input_file=None
        output_file=None
        check_is_natural(sys.argv[1])
        k=int(sys.argv[1])
        max_iter=200
        if(len(sys.argv)==4):
            input_file=sys.argv[2]
            output_file=sys.argv[3]
            
        elif (len(sys.argv)==5):
            check_is_natural(sys.argv[2])
            max_iter=int(sys.argv[2])
            input_file=sys.argv[3]
            output_file=sys.argv[4]
        else: 
            invalid_input()
    
        #Reading from file
        data_point_matrix=file_to_matrix(input_file)
        if(k > len(data_point_matrix)):
            invalid_input()
        centroids=kmeans(data_point_matrix,k, max_iter)
        write_to_file(output_file,centroids)
    except Exception as e:
        print("An Error Has Occurred")
        exit()

main()